### PR TITLE
feat Device commands refactoring

### PR DESCRIPTION
**Pull request by bflorian**
Refactoring of PR #74 to incorporate feedback

> Could we create a SelectingInputAPICommand and use that rather than having this use buildTableOutput like this? That would also avoid the extraneous output flags.

Done

> I would rather see this [commands] asked as multiple questions. If the user has to put it all in one line, why not just have them enter it on the command line?

Done, see below

> I think this [component-status] also would benefit from multiple questions.

Done, see below

> I think it would be easier to read the code if we created a couple of extra specific-purpose "Id" types in the command classes instead (that would then have named properties like "capability" or "component").

Pass-through of arguments was removed entirely

#### command line
```
$ st devices:component-status aa705336-bea7-4d6a-924b-c2fc313c6ad4 outlet1
┌────────────┬───────────┬───────┐
│ Capability │ Attribute │ Value │
├────────────┼───────────┼───────┤
│ switch     │ switch    │ "off" │
└────────────┴───────────┴───────┘
```
#### interactive one component
```
$ st devices:component-status 11
┌─────────────┬──────────────────────────┬───────────────────────────────────────────┐
│ Capability  │ Attribute                │ Value                                     │
├─────────────┼──────────────────────────┼───────────────────────────────────────────┤
│ healthCheck │ checkInterval            │                                           │
│ healthCheck │ healthStatus             │                                           │
│ healthCheck │ DeviceWatch-Enroll       │ {"protocol":"cloud","scheme":"UNTRACKED"} │
│ healthCheck │ DeviceWatch-DeviceStatus │ "online"                                  │
│ switchLevel │ level                    │ 100 %                                     │
│ switch      │ switch                   │ "off"                                     │
└─────────────┴──────────────────────────┴───────────────────────────────────────────┘
```

#### interactive multiple components
```
$ st devices:component-status 60

Components:
┌───┬────────────────┐
│ 1 │ main (default) │
│ 2 │ outlet1        │
│ 3 │ outlet2        │
└───┴────────────────┘
? Enter component id 2
┌────────────┬───────────┬───────┐
│ Capability │ Attribute │ Value │
├────────────┼───────────┼───────┤
│ switch     │ switch    │ "off" │
└────────────┴───────────┴───────┘
```

#### 

#### command line
```
$ st devices:capability-status f1360307-d7fc-4ff6-b092-2569845f00ab main colorTemperature
┌──────────────────┬────────┐
│ Attribute        │ Value  │
├──────────────────┼────────┤
│ colorTemperature │ 2900 K │
└──────────────────┴────────┘
```

#### interactive
```
$ st devices:capability-status
┌────┬───────────────────────────────┬────────────────────────────────┬──────────────┬──────────────────────────────────────┐
│ #  │ Label                         │ Name                           │ Type         │ Device Id                            │
├────┼───────────────────────────────┼────────────────────────────────┼──────────────┼──────────────────────────────────────┤
│ 1  │ Bathroom Fan                  │ c2c-switch                     │ VIPER        │ 59bad3c5-08e6-421d-abe9-c075a6085c13 │
│ 2  │ Bathroom Leak Detector        │ Virtual STS Leak Detector      │ VIPER        │ f0556074-caca-49df-b524-e441c3cd57f4 │
│ 3  │ Bathroom Light                │ c2c-dimmer                     │ VIPER        │ a7c475bb-f9ad-48b8-a6c5-60c2e33096a3 │
│ 4  │ Floor Lamp                    │ c2c-rgbw-color-bulb            │ VIPER        │ 84abbe1b-e347-4be4-879a-622800c71e0f │
│ 5  │ Two Channel Outlet            │ VirtualSTS Two Channel Outlet  │ VIPER        │ 02ade346-b533-48ab-923d-23e99a7c837f │
└────┴───────────────────────────────┴────────────────────────────────┴──────────────┴──────────────────────────────────────┘
? Enter id or index 5

Components:
┌───┬────────────────┐
│ 1 │ main (default) │
│ 2 │ outlet1        │
│ 3 │ outlet2        │
└───┴────────────────┘
? Enter component id main

Capabilities:
┌───┬──────────────────┐
│ 1 │ switch (default) │
│ 2 │ healthCheck      │
└───┴──────────────────┘
? Enter capability id switch
┌───────────┬───────┐
│ Attribute │ Value │
├───────────┼───────┤
│ switch    │ "off" │
└───────────┴───────┘
```

#### command line input
```
$ st devices:commands aa705336-bea7-4d6a-924b-c2fc313c6ad4 switch:on
Command executed successfully
```
```
$ st devices:commands f1360307-d7fc-4ff6-b092-2569845f00ab 'switchLevel:setLevel(30)'
Command executed successfully
```

#### command line file input
```
$ st devices:commands 5e8422f8-9994-46a9-944e-70cc71b55ed8 -i cmd.yml
Command executed successfully
```
`cmd.yml` file:
```
- component: outlet2
  capability: switch
  command: on
```

#### interactive single component
```
$ st devices:commands 9

Living Room Lamp

Capabilities:
┌───┬──────────────────┐
│ 1 │ switch           │
│ 2 │ switchLevel      │
│ 3 │ colorControl     │
│ 4 │ colorTemperature │
│ 5 │ refresh          │
│ 6 │ healthCheck      │
└───┴──────────────────┘
? Enter capability index or id 4

Commands:
┌───┬───────────────────────────────────────────┐
│ 1 │ setColorTemperature(temperature<integer>) │
└───┴───────────────────────────────────────────┘
? Enter command 1
? Enter command arguments (temperature) 2950
Command executed successfully
```
#### interactive multi-component
```
$ st devices:commands 60

Kitchen Outlet

Components:
┌───┬─────────┐
│ 1 │ main    │
│ 2 │ outlet1 │
│ 3 │ outlet2 │
└───┴─────────┘
? Enter component index or id 1

Capabilities:
┌───┬─────────────┐
│ 1 │ switch      │
│ 2 │ healthCheck │
└───┴─────────────┘
? Enter capability index or id 1

Commands:
┌───┬───────┐
│ 1 │ off() │
│ 2 │ on()  │
└───┴───────┘
? Enter command 2
Command executed successfully
```
#### interactive shortcut
```
$ st devices:commands 60

Kitchen Outlet

Components:
┌───┬─────────┐
│ 1 │ main    │
│ 2 │ outlet1 │
│ 3 │ outlet2 │
└───┴─────────┘
? Enter component index or id outlet1:switch:off()
Command executed successfully
```

### CODE REVIEW

File: packages/cli/package.json [hunk 1]
```diff
@@ -66,6 +66,7 @@
 		"@types/cli-table": "^0.3.0",
 		"@types/lodash": "^4.14.155",
 		"@types/yeoman-environment": "^2.3.3",
+		"jest": "^26.1.0",
```
[the remark this answers is not in the document]
Comment (bflorian) [reply packages/cli/package.json:69]: How should I be running the tests then? When I run `npm run test` in the `cli` directory I get this:
```
cli (device-commands-take2)$ npm run test

> @smartthings/cli@0.0.0 test /Users/bobflorian/st/pi/smartthings-cli/packages/cli
> jest

sh: jest: command not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! @smartthings/cli@0.0.0 test: `jest`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the @smartthings/cli@0.0.0 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

That's immediately after a full clean and bootstrap

File: packages/cli/src/commands/devices/update.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+	static args = [
+		{
+			name: 'id',
+			description: 'the device id',
+		},
+		{
+			name: 'label',
+			description: 'the new device label',
+		},
```
[the remark this answers is not in the document]
Comment (bflorian) [reply packages/cli/src/commands/devices/update.ts]: Removed

File: packages/cli/src/commands/installedapps/rename.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	static flags = SelectingOutputAPICommand.flags
+
+	static args = [
+		{
+			name: 'id',
+			description: 'the device id',
+		},
+		{
+			name: 'name',
+			description: 'the new device name',
```
[the remark this answers is not in the document]
Comment (bflorian) [reply packages/cli/src/commands/installedapps/rename.ts]: Fixed

File: packages/lib/src/io-command.ts [hunk 4]
```diff
@@ -583,29 +592,47 @@ export abstract class SelectingAPICommand<L> extends SelectingAPICommandBase<str
...
 	protected abstract async getIdFromUser(items: L[]): Promise<ID>
 
-	protected async processNormally(id: ID | undefined,
+	protected async translateToId(idOrIndex: ID | string, listFunction: ListCallback<L>): Promise<ID> {
+		// @ts-ignore
+		return idOrIndex
+	}
+
+	protected acceptIndexId = false
+	protected inputPrompt = 'Enter id or index'
```
[the remark this answers is not in the document]
Comment (bflorian) [reply packages/lib/src/io-command.ts]: Fixed

File: packages/lib/src/io-command.ts [hunk 5]
```diff
@@ -583,29 +592,47 @@ export abstract class SelectingAPICommand<L> extends SelectingAPICommandBase<str
...
 
 export abstract class SelectingInputOutputAPICommandBase<ID, I, O, L> extends APICommand {
+	private _entityId?: ID
 	protected abstract async getIdFromUser(items: L[]): Promise<ID>
 
-	protected async processNormally(id: ID | undefined,
+	protected async translateToId(idOrIndex: ID | string, listFunction: ListCallback<L>): Promise<ID> {
+		// @ts-ignore
+		return idOrIndex
+	}
```
[the remark this answers is not in the document]
Comment (bflorian) [reply packages/lib/src/io-command.ts]: Done (in three places)

File: packages/cli/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (rossiam) [reply packages/cli/package.json:69]: You need to _everything_ in the CLI through lerna. To run all the tests run `lerna run test` in the root directory. If you want to run only tests for the CLI, you can tell lerna to run the command only in the "cli" package like this:

```
lerna run test  --scope @smartthings/cli
```

You can also use `lerna exec ...` to run "jest" directly, if you, for example, want to watch tests:

```
lerna exec --scope @smartthings/cli "jest --watch"
```

**Pull request merged**